python/revng/daemon/event_manager.py

```diff
@@ -0,0 +1,135 @@
+        for fifo_definition in fifo_string.split(","):
+            fifo_path, string_type = fifo_definition.split(":", 1)
+            if string_type == "begin":
+                self.add_event_listener(EventType.BEGIN, BeginListener(Path(fifo_path)))
+            elif string_type == "context":
+                self.add_event_listener(EventType.CONTEXT, ContextListener(Path(fifo_path)))
+            else:
+                logging.warn(f"Unknown event type specified for fifo: {string_type}")
+
+    def add_event_listener(self, type_: EventType, callback: EventHandler):
+        self.listeners[type_].append(callback)
+
+    def handle_event(self, type_: EventType, headers: Headers):
+        if type_ not in self.listeners:
+            return
+
+        def handler():
+            for listener in self.listeners[type_]:
+                listener(self.manager, type_, self.get_token(headers))
+
+        # Run the callbacks in an executor to avoid blocking the result
+        result = self.executor.submit(handler)
+        result.add_done_callback(self.handle_event_result)
+
+    @staticmethod
+    def get_token(headers: Headers) -> str:
+        if "authorization" not in headers or not headers["authorization"].startswith("Bearer "):
+            return ""
+        token = headers["authorization"][len("Bearer ") :]
+        if re.search(r"\s", token):
+            return ""
+        return token
+
+    @staticmethod
+    def handle_event_result(result_future: Future):
+        ex = result_future.exception()
+        if ex is not None:
+            logging.error("Exception raised while handling event callbacks", exc_info=ex)
+            signal.raise_signal(signal.SIGINT)
+
+
+class BaseListener(ABC, EventHandler):
+    def __init__(self, fifo_path: Path):
+        self.fifo = open(fifo_path, "w")  # noqa: SIM115
+
+    @abstractmethod
+    def __call__(self, manager: Manager, type_: EventType, token: str):
+        pass
+
+    def write_fifo(self, content: str):
+        try:
+            self.fifo.write(content)
+            self.fifo.flush()
+        except (BrokenPipeError, FileNotFoundError, PermissionError, TimeoutError) as e:
+            logging.warn(
+                f"Encountered error when writing to notification pipe: {repr(e)}",
+            )
+
+    def __del__(self):
+        self.fifo.close()
+
+
+class BeginListener(BaseListener):
+    def __call__(self, manager: Manager, type_: EventType, token: str):
+        assert type_ == EventType.BEGIN
+        begin_step = manager.get_step("begin")
+        assert begin_step is not None
+
+        tmpdir = mkdtemp()
+        begin_step.save(tmpdir)
+        self.write_fifo(f"PUSH begin {tmpdir} {token}\n")
+
+
+class ContextListener(BaseListener):
+    def __call__(self, manager: Manager, type_: EventType, token: str):
+        assert type_ == EventType.CONTEXT
+        tmpdir = mkdtemp()
+        manager.save_context(tmpdir)
+        self.write_fifo(f"PUSH context {tmpdir} {token}\n")
```

Comment:
Is it really necessary to have two of them? Let's merge them.

Reply:
Eh, I'd say it'd only make it harder and more complicated to read if they were merged
Also this snippet makes it pretty explicit what they do (and it's one and one thing only)
```python
            if string_type == "begin":
                self.add_event_listener(EventType.BEGIN, BeginListener(Path(fifo_path)))
            elif string_type == "context":
                self.add_event_listener(EventType.CONTEXT, ContextListener(Path(fifo_path)))
```

Reply:
To me it seems a lot of boiler plate to add a new event for little benefit.

> more complicated to read

Why?

Reply:
Because the resulting class would look like
```python
class Listener:
  def __init__(self, listen_type: Begin | Context):
    self.listen_type = listen_type

  def __call__(self, ...):
    if self.listen_type == Begin:
      <do begin stuff>
    if self.listen_type == Context:
      <do context stuff>
 ```
 
 Instead here you know that each class does exactly one kind of operation

Reply:
As an implementor of a listener I just want to write a function that takes `Begin`/`Context` as an argument and act based on that. I don't really care about the rest.

Can you figure out enough infrastructure to make this possible?

Reply:
aaaah *massages nipples*